### PR TITLE
New API Guide and API Cleanup

### LINKED ISSUE #156: API calls need to return startobs/limit/result_count metadata with each response
Originally reported by: **lisa ballard (Bitbucket: [basilleaf](https://bitbucket.org/basilleaf), GitHub: [basilleaf](https://github.com/basilleaf))**

---

when a large result is returned there should be a field that describes the total result count, current limit, and page. This is common in json APIs around the web and now I see why it is so helpful.. 

Here is background on how this came up:

Begin forwarded message:
On May 22, 2015, at 12:01 PM, Michael Aye wrote:

Hi Lisa,

how are you?

I am noticing an upper limit of 100 results for the API requests?

This one for example delivers 2261 results on the website, but the API keeps telling me 100:

'http://pds-rings-tools.seti.org/opus/api/files.json?instrumentid=Cassini+ISS&target=S+RINGS&projectedradialresolution1=0.5&projectedradialresolution2=2'

Did you have a limit put in for testing?

Michael

From: Lisa Ballard lballard@seti.org
Date: May 22, 2015 at 12:30:40 PM PDT
To: Michael Aye <>

Hi Michael! 

You can control how many results are returned by setting the limit variable in the url, so for example here at the end of the url I’ve added ‘&limit=3000' to assure we get all results for this particular search:

http://pds-rings-tools.seti.org/opus/api/files.json?instrumentid=Cassini+ISS&target=S+RINGS&projectedradialresolution1=0.5&projectedradialresolution2=2&limit=3000

The API only returns 100 results at a time, ‘pages’ if you will. You can navigate the pages by setting the page variable in the url, like here I’ve added page=5 at the end.

http://pds-rings-tools.seti.org/opus/api/files.json?instrumentid=Cassini+ISS&target=S+RINGS&projectedradialresolution1=0.5&projectedradialresolution2=2&page=5

This behavior is actually by design, as some large results will spin for awhile to serve thousands of results.. but I think since you asked it could be made clearer what is happening in the response itself. For example there could be another variable besides ‘data’ that tells you things like what page you are on, what the current limit is set to, and what the total result count is.. so I’m going to add this to our issue tracker for a future improvement. 

hope this helps!  
-lisa

---
- Bitbucket: https://bitbucket.org/ringsnode/opus2/issue/156

### LINKED ISSUE #495: Need V2 of files API that uses slug names instead of pretty names
Currently the files API returns the "pretty name" for each file (like "Browse Image (full-size)". We should actually return the slug name ("browse-full"), so that we have the freedom to change the pretty name in the future.

### LINKED ISSUE #647: Results API: Cleanup needed
The following APIs need internal clean to get rid of response_formats(). Most have broken returns for CSV and/or ZIP as well.

api_get_images_by_size
api_get_images
api_get_image

### LINKED ISSUE #915: Need new API Guide
The current API Guide is based on a horrible YAML infrastructure that limits the kind of formatting we can do. For the February goal we need a new, pretty, and complete API Guide.

### LINKED ISSUE #942: Need API call that returns all possible product types
This was originally requested by Michael Aye.

api/files.json allows a list of product types, but there's no way to know what the valid product types are so that you can provide them.

### PR DESCRIPTION
**Pull request by rfrenchseti**
- Fixes #915 (Need new API Guide). The API Guide has been completely rewritten. It now uses Markdown instead of the weird YAML thing we used to use.
- Fixes #942 (Need API call that returns all possible product types). There are new API calls `/api/product_types.json` and `/api/product_types/[opusid].json`.
- Fixes #647 (Results API: Cleanup needed). `api_get_images_by_size`, `api_get_images`, `api_get_image` have been cleaned up. No API calls use the old `response_formats` function and it has been deleted. Misc other API calls have been cleaned up and in some cases the returned data has been modified to make it more consistent.
- Fixes #495 (Need V2 of files API that uses slug names instead of pretty names). The existing `/api/files` APIs have been changed to return the short name instead of the pretty name of product types. I did this instead of creating a new V2 version of the API because no one was using the files API anyway so it wouldn't break anything.
- Fixes #156 (API calls need to return startobs/limit/result_count metadata with each response). There is a new `available` field in some API returns that gives the result count in addition to the existing requested `limit` and actually returned `count`.
- There is a new feature that allows creation and download of PDFs for many of the help selections. It isn't available for "Citing OPUS" because the converter doesn't handle the QR code graphics properly.
- I removed the `tutorial` help API because we don't use it and don't plan to.

**NOTE**

- You must `pip install -r requirements-python3.txt` (or the appropriate Mac or Windows versions) to update your dependencies.
- In order for PDF conversion to work, you must install `wkhtmltopdf` from https://wkhtmltopdf.org/  You must install a version with the Qt patches.
- The variable `STATIC_ROOT` in `opus_secrets.py` needs to be renamed to `OPUS_STATIC_ROOT`.
- Running tests on this branch requires the use of the new OPUS test database `opus3_test_200103`. The two big changes in this database are that all HST OPUSIDs have 9 characters, and all download product types use "_" instead of "-" in their short names.

### CODE REVIEW

**Comment by rfrenchseti** [comment]
Since you may not be able to create the PDF files on your own, here is what they look like:

[opus_api_guide.pdf](https://github.com/SETI/pds-opus/files/4024500/opus_api_guide.pdf)
[opus_volumes.pdf](https://github.com/SETI/pds-opus/files/4024501/opus_volumes.pdf)
[opus_faq.pdf](https://github.com/SETI/pds-opus/files/4024502/opus_faq.pdf)
[opus_getting_started.pdf](https://github.com/SETI/pds-opus/files/4024503/opus_getting_started.pdf)
[opus_about.pdf](https://github.com/SETI/pds-opus/files/4024504/opus_about.pdf)

**Comment by rfrenchseti** [comment]
I made a mistake with renaming `STATIC_ROOT`. There are now two variables, `STATIC_ROOT` and `OPUS_STATIC_ROOT`. The first is ignored in a non-production environment (e.g. runserver). The second should be a pointer inside the OPUS source tree in a non-production environment, or the really static location in a production environment. See the comments in opus_template.py.

File: opus/application/apps/help/templates/help/about.html [hunk 1]
```diff
@@ -39,7 +39,7 @@
...
+			<a href="https://tools.pds-rings.seti.org/opus/#/SURFACEGEOplutocenterresolution2=5.00000&SURFACEGEOplutophase1=160.000&surfacegeometrytargetname=Pluto&qtype-SURFACEGEOplutophase=any&cols=opusid,instrument,time1,SURFACEGEOplutoplanetographiclatitude1,SURFACEGEOplutoplanetographiclatitude2,SURFACEGEOplutoIAUwestlongitude1,SURFACEGEOplutoIAUwestlongitude2,SURFACEGEOplutocenterdistance,SURFACEGEOplutocenterresolution,SURFACEGEOplutophase1,SURFACEGEOplutophase2&widgets=SURFACEGEOplutocenterresolution,SURFACEGEOplutophase,surfacegeometrytargetname&order=time1,opusid&view=search&browse=gallery&cart_browse=gallery&startobs=1&cart_startobs=1&detail=" target="_blank">this query</a> searches for:
```
Comment (juzen2003) [inline opus/application/apps/help/templates/help/about.html]: Need to update this later once pull https://github.com/SETI/pds-opus/pull/965 is merged.

File: opus/application/apps/help/api_guide.md [hunk 2]
```diff
@@ -0,0 +1,2028 @@
...
+| `page` | The observation data |
+
+`page` is a list with one entry per returned observation. Each entry is itself a list, with one entry per requested metadata field, in the same order as was requested with `cols`.
+
+Example:
+
+* Retrieve data in JSON format for the first three Cassini ISS images that contain Enceladus' south pole (latitude 70 degrees or greater) and have a phase angle at Enceladus of 160 degrees or greater.
+
+    %EXTLINK%%HOST%/opus/api/data.json?instrument=Cassini+ISS&SURFACEGEOenceladusplanetographiclatitude1=70&SURFACEGEOenceladuscenterphaseangle1=160&order=time1&cols=opusid,target,time1,SURFACEGEOenceladuscenterphaseangle&startobs=5&limit=3%ENDEXTLINK%
+
```
Comment (juzen2003) [inline opus/application/apps/help/api_guide.md:275]: For record, we have to update this later once pull #965 is merged.

File: opus/application/apps/help/api_guide.md [hunk 3]
```diff
@@ -0,0 +1,2028 @@
...
+#### CSV Return Format
+
+The return value is a series of text lines. The first line contains the names of the requested metadata fields. After that is one line per observation containing the requested metadata.
+
+Example:
+
+* Retrieve data in CSV format for the first three Cassini ISS images that contain Enceladus' south pole (latitude 70 degrees or greater) and have a phase angle at Enceladus of 160 degrees or greater.
+
+    %EXTLINK%%HOST%/opus/api/data.csv?instrument=Cassini+ISS&SURFACEGEOenceladusplanetographiclatitude1=70&SURFACEGEOenceladuscenterphaseangle1=160&order=time1&cols=opusid,target,time1,SURFACEGEOenceladuscenterphaseangle&startobs=5&limit=3%ENDEXTLINK%
+
```
Comment (juzen2003) [inline opus/application/apps/help/api_guide.md:323]: Same here, might want to update the URL later.

File: opus/application/apps/help/api_guide.md [hunk 4]
```diff
@@ -0,0 +1,2028 @@
...
+#### HTML Return Format
+
+The return value is an HTML table. The table header contains the names of the requested metadata fields. The table rows contain the requested metadata.
+
+Example:
+
+* Retrieve data in HTML format for the first three Cassini ISS images that contain Enceladus' south pole (latitude 70 degrees or greater) and have a phase angle at Enceladus of 160 degrees or greater.
+
+    %EXTLINK%%HOST%/opus/api/data.html?instrument=Cassini+ISS&SURFACEGEOenceladusplanetographiclatitude1=70&SURFACEGEOenceladuscenterphaseangle1=160&order=time1&cols=opusid,target,time1,SURFACEGEOenceladuscenterphaseangle&startobs=5&limit=3%ENDEXTLINK%
+
```
Comment (juzen2003) [inline opus/application/apps/help/api_guide.md:342]: Same here, might want to update the URL later.

File: opus/application/apps/help/api_guide.md [hunk 5]
```diff
@@ -0,0 +1,2028 @@
...
+    <ul>
+    <li>
+    <img id="med__co-iss-n1460962327" src="https://pds-rings.seti.org/holdings/previews/COISS_2xxx/COISS_2002/data/1460960653_1461048959/N1460962327_1_med.jpg">
+    </li>
+    <li>
+    <img id="med__co-iss-n1460962415" src="https://pds-rings.seti.org/holdings/previews/COISS_2xxx/COISS_2002/data/1460960653_1461048959/N1460962415_1_med.jpg">
+    </li>
+    </ul>
+%ENDCLASS%
+
```
Comment (juzen2003) [inline opus/application/apps/help/api_guide.md:1157]: I tried to run this link but didn't get any returned images: https://tools.pds-rings.seti.org/opus/api/images/med.html?volumeid=COISS_2002&startobs=10&limit=2, and if I ran through the test database, there will be 2 images: http://localhost:8000/opus/api/images/med.html?volumeid=COISS_2002&startobs=10&limit=2

File: opus/application/apps/help/api_guide.md [hunk 6]
```diff
@@ -0,0 +1,2028 @@
...
+%CODE%
+[
+  {
+    "product_type": "coiss_raw",
+    "description": "Raw image"
+  },
+  {
+    "product_type": "coiss_calib",
+    "description": "Calibrated im
```
Comment (juzen2003) [inline opus/application/apps/help/api_guide.md:1856]: Return value with SURFACEGEO needs to be updated after merging with #965.

Comment (juzen2003) [inline opus/application/apps/help/api_guide.md:1899]: Return value with SURFACEGEO needs to be updated after merging with #965.

Comment (juzen2003) [inline opus/application/apps/help/api_guide.md:2010]: Return value with SURFACEGEO needs to be updated after merging with #965.

Comment (juzen2003) [inline opus/application/apps/help/api_guide.md:1987]: SURFACEGEO slug needs to be updated after merging with #965.

File: opus/application/apps/help/templates/help/apiguide.html [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        	    {% for v in fields %}
+                    <tr>
+                        <td>{{ v.category }}</td>
+                        <td>{{ v.label }}
+                            {% if v.pretty_units %}
+                                <br/>{{ v.pretty_units }}
+                            {% endif %}
+                        </td>
+                        <td>{{ v.slug }}</td>
+                    </tr>
```
Comment (juzen2003) [inline opus/application/apps/help/templates/help/apiguide.html:25]: Indentations are a bit off for the tr tag

File: opus/application/apps/help/templates/help/apiguide.html [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    <div id="field_data" class="table-responsive">
+        <table class="table table-sm table-striped table-hover">
+            <thead class="thead-dark">
+                <tr>
+                    <th>Category<br/>&nbsp;</th>
+                    <th>Label<br/>Units</th>
+                    <th>Slug<br/>&nbsp;</th>
+                </tr>
+            </thead>
+    		<tbody>
```
Comment (juzen2003) [inline opus/application/apps/help/templates/help/apiguide.html]: I think we don't need the indentation here.

File: opus/application/apps/help/templates/help/apiguide_print.html [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    <h1 id="availablefields" class="op-new-page">Available Metadata Fields</h1>
+    <table class="table table-sm table-striped">
+        <thead class="thead-dark">
+            <tr>
+                <th>Category<br/>&nbsp;</th>
+                <th>Label<br/>Units</th>
+                <th>Slug<br/>&nbsp;</th>
+            </tr>
+        </thead>
+		<tbody>
```
Comment (juzen2003) [inline opus/application/apps/help/templates/help/apiguide_print.html]: I think we don't need the indentation here.

File: opus/application/apps/help/templates/help/apiguide_print.html [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+                    <td>{{ v.category }}</td>
+                    <td>{{ v.label }}
+                        {% if v.pretty_units %}
+                            <br/>{{ v.pretty_units }}
+                        {% endif %}
+                    </td>
+                    <td>{{ v.slug }}</td>
+                </tr>
+    	    {% endfor %}
+		</tbody>
```
Comment (juzen2003) [inline opus/application/apps/help/templates/help/apiguide_print.html]: Indentations are off for tbody and tr tags.

File: opus/application/apps/results/templates/results/image_list.html [hunk 11]
```diff
@@ -1,7 +1,7 @@
...
-	<li>
-		<img id="{{ size }}__{{ i.opus_id }}" src="{{ i.url }}">
-  </li>
-{% endfor %}
-</body>
+<ul>
+    {% for i in data %}
+    	<li>
+    		<img id="{{ size }}__{{ i.opus_id }}" src="{{ i.url }}">
+        </li>
```
Comment (juzen2003) [inline opus/application/apps/results/templates/results/image_list.html:5]: li and /li not aligned.

File: opus/application/static_media/css/opus.css [hunk 12]
```diff
@@ -1883,6 +1883,81 @@ footer {
...
+    word-break: break-all !important;
+}
+
+.op-help-api-guide .op-api-guide-code code {
+    white-space: pre-wrap !important;
+    word-break: break-all !important;
+}
+
+.op-help-api-guide .op-table-nonfluid {
+   width: auto !important;
```
Comment (juzen2003) [inline opus/application/static_media/css/opus.css]: Missing one space.

File: opus/application/static_media/js/opus.js [hunk 13]
```diff
@@ -849,34 +849,42 @@ var opus = {
          * Given the name of a help menu entry, open the help pane and load the
          * help contents.
          */
-        let url = "/opus/__help/";
+        let base_url = "/opus/__help/";
+        let pdf_url = false;
```
Comment (juzen2003) [inline opus/application/static_media/js/opus.js]: Maybe we want to use camel case when declaring variables in js code?

File: opus/application/static_media/js/opus.js [hunk 14]
```diff
@@ -923,21 +936,27 @@ var opus = {
...
+                    newTabWindow.document.getElementsByTagName("title")[0].innerHTML = header;
                     $(newTabWindow.document.body).append(contents)
                         .css({
                             overflow: "auto",
                             margin: "1.5em",
                             backgroundColor: "inherit"
                         });
                 });
+                $(".op-open-help-pdf").on("click", function(e) {
+                    let pdf_url = $(".op-open-help-pdf").data("action");
```
Comment (juzen2003) [inline opus/application/static_media/js/opus.js]: Maybe we want to use camel case here?

**Review by juzen2003 — CHANGES_REQUESTED** [review]
- Database used: opus3_test_200107
- All Django tests pass: Y
- JSHINT run on all affected files: Y (opus.js & search.js)
- Tested on Chrome, Firefox, Safari, and iOS: Y
  (Remember to test all browser sizes)
- Tested for race conditions (with delayed API calls if applicable): NA (run each api one at a time)

- Comment:
  - For the indentation issues in html, they look matched in atom but not in github file display. That's because regular 4 spaces look like 4 "." in atom, but the places with indentation issues have something like ">>" in atom.

File: opus/application/apps/help/templates/help/about.html (hunk 1 above)
Comment (rfrenchseti) [reply opus/application/apps/help/templates/help/about.html]: Fixed

File: opus/application/apps/help/api_guide.md (hunk 2 above)
Comment (rfrenchseti) [reply opus/application/apps/help/api_guide.md:275]: Fixed

File: opus/application/apps/help/api_guide.md (hunk 3 above)
Comment (rfrenchseti) [reply opus/application/apps/help/api_guide.md:323]: Fixed

File: opus/application/apps/help/api_guide.md (hunk 4 above)
Comment (rfrenchseti) [reply opus/application/apps/help/api_guide.md:342]: Fixed

File: opus/application/apps/help/api_guide.md (hunk 5 above)
Comment (rfrenchseti) [reply opus/application/apps/help/api_guide.md:1157]: Yes there was a bug in the current implementation and I fixed it.

File: opus/application/apps/help/api_guide.md (hunk 6 above)
Comment (rfrenchseti) [reply opus/application/apps/help/api_guide.md:1856]: Fixed

Comment (rfrenchseti) [reply opus/application/apps/help/api_guide.md:1899]: Fixed

Comment (rfrenchseti) [reply opus/application/apps/help/api_guide.md:2010]: Fixed

Comment (rfrenchseti) [reply opus/application/apps/help/api_guide.md:1987]: Fixed

File: opus/application/apps/help/templates/help/apiguide.html (hunk 7 above)
Comment (rfrenchseti) [reply opus/application/apps/help/templates/help/apiguide.html:25]: Fixed

File: opus/application/apps/help/templates/help/apiguide.html (hunk 8 above)
Comment (rfrenchseti) [reply opus/application/apps/help/templates/help/apiguide.html]: Fixed

File: opus/application/apps/help/templates/help/apiguide_print.html (hunk 9 above)
Comment (rfrenchseti) [reply opus/application/apps/help/templates/help/apiguide_print.html]: Fixed

File: opus/application/apps/help/templates/help/apiguide_print.html (hunk 10 above)
Comment (rfrenchseti) [reply opus/application/apps/help/templates/help/apiguide_print.html]: Fixed

File: opus/application/apps/results/templates/results/image_list.html (hunk 11 above)
Comment (rfrenchseti) [reply opus/application/apps/results/templates/results/image_list.html:5]: Fixed

File: opus/application/static_media/css/opus.css (hunk 12 above)
Comment (rfrenchseti) [reply opus/application/static_media/css/opus.css]: Fixed

File: opus/application/static_media/js/opus.js (hunk 13 above)
Comment (rfrenchseti) [reply opus/application/static_media/js/opus.js]: Fixed

File: opus/application/static_media/js/opus.js (hunk 14 above)
Comment (rfrenchseti) [reply opus/application/static_media/js/opus.js]: Fixed

**Review by rfrenchseti — COMMENTED** [review]
Changes made as requested

**Review by juzen2003 — APPROVED** [review]

**Pull request merged**